File: content/snippets/cpp/interview_lcci/17/20_continuous_median.cpp
#include <queue>
using namespace std;

class MedianFinder {
    priority_queue<int> lo;                          // max-heap
    priority_queue<int, vector<int>, greater<int>> hi; // min-heap
public:
    void addNum(int num) {
        lo.push(num);
        hi.push(lo.top()); lo.pop();
        if (hi.size() > lo.size()) {
            lo.push(hi.top()); hi.pop();
        }
    }

    double findMedian() {
        if (lo.size() > hi.size()) return lo.top();
        return (lo.top() + hi.top()) / 2.0;
    }
};

File: content/snippets/scala3/interview_lcci/17/20_continuous_median.scala
import scala.collection.mutable

class MedianFinder:
  // lo: max-heap, hi: min-heap
  private val lo = mutable.PriorityQueue[Int]()        // max-heap
  private val hi = mutable.PriorityQueue[Int]()(Ordering[Int].reverse) // min-heap

  def addNum(num: Int): Unit =
    lo.enqueue(num)
    hi.enqueue(lo.dequeue())
    if (hi.size > lo.size) lo.enqueue(hi.dequeue())

  def findMedian(): Double =
    if (lo.size > hi.size) lo.head.toDouble
    else (lo.head + hi.head) / 2.0
